source/slang/slang-check-type.cpp

```diff
@@ -305,14 +306,36 @@ namespace Slang
                     }
                     // TODO: this is one place where syntax should get cloned!
                     if (outProperType)
-                        args.add(valParam->initExpr);
+                        args.add(ExtractGenericArgVal(valParam->initExpr));
+                }
+                else if (auto constraintParam = as<GenericTypeConstraintDecl>(member))
+                {
+                    auto genericParam = as<DeclRefType>(constraintParam->sub.type)->getDeclRef();
+                    if (!genericParam)
+                        return false;
+                    auto genericTypeParamDecl = as<GenericTypeParamDecl>(genericParam.getDecl());
+                    if (!genericTypeParamDecl)
+                        return false;
+                    auto defaultType = CheckProperType(genericTypeParamDecl->initType);
+                    auto witness = tryGetSubtypeWitness(defaultType, CheckProperType(constraintParam->sup));
+                    if (!witness)
+                    {
+                        // diagnose
+                        SLANG_ABORT_COMPILATION("unexpected");
```

Comment:
Don't use SLANG_ABORT_COMPILATION, let's produce a proper diagnostic here.
Use SLANG_UNEXPECTED instead of SLANG_ABORT_COMPILATION.